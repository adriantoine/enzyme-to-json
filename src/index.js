import omit from 'lodash.omit';
import pickBy from 'lodash.pickby';
import compact from 'lodash.compact';
import get from 'lodash.get';

export function shallowToJson(wrapper) {
    const type = wrapper.type();
    if (!type) {
        return wrapper.node;
    }

    const children = compact(wrapper.children().map(c => shallowToJson(c)));
    const props = pickBy(wrapper.props(), val => val !== undefined);

    const json = {
        type: wrapper.name(),
        props: omit(props, 'children'),
        children: (children.length) ? children : null,
        $$typeof: Symbol.for('react.test.json'),
    };

    // If the type of element is not a function, it means that is a DOM element
    if (typeof type !== 'function') {
        return json;
    }

    // We need to get the rendered component in Enzyme internals
    // because it won't be returned by `.children()`
    const element = get(wrapper, 'node._reactInternalInstance._renderedComponent._currentElement');
    if(!element) {
        return json;
    }

    const jsonChildren = json.children;
    json.children = [{
        type: element.type,
        props: omit(element.props, 'children'),
        children: jsonChildren,
        $$typeof: Symbol.for('react.test.json'),
    }];

    return json;
}

export {shallowToJson as mountToJson};

const renderChildToJson = child => {
    if(!child) return;

    if(child.type === 'tag') {
        return {
            type: child.name,
            props: child.attribs,
            children: child.children && child.children.length ? compact(child.children.map(renderChildToJson)) : null,
            $$typeof: Symbol.for('react.test.json'),
        };
    } else if(child.type === 'text') {
        return child.data;
    }
};
export const renderToJson = wrapper =>
    renderChildToJson(wrapper.children()[0]);
