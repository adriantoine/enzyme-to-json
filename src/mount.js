import compact from 'lodash.compact';
import omitBy from 'lodash.omitby';
import values from 'object-values';
import {isDOMComponent, isElement} from 'enzyme/build/react-compat';
import {internalInstance, propsOfNode} from 'enzyme/build/Utils';
import {typeName} from 'enzyme/build/Debug';
import {childrenOfNode} from 'enzyme/build/ShallowTraversal';

function instToJson(inst) {
    if (typeof inst === 'string' || typeof inst === 'number') return inst;
    if (!inst) return '';

    if (!inst.getPublicInstance) {
        const internal = internalInstance(inst);
        return instToJson(internal);
    }
    const publicInst = inst.getPublicInstance();

    if (typeof publicInst === 'string' || typeof publicInst === 'number') return publicInst;
    if (!publicInst && !inst._renderedComponent) return '';

    const currentElement = inst._currentElement;
    const type = typeName(currentElement);
    const props = omitBy(propsOfNode(currentElement), (val, key) => key === 'children' || val === undefined);
    const children = [];
    if (isDOMComponent(publicInst)) {
        const renderedChildren = inst._renderedChildren;
        if (!renderedChildren) {
            children.push(...childrenOfNode(currentElement));
        } else {
            children.push(...values(renderedChildren));
        }
    } else if (
        isElement(currentElement) &&
        typeof currentElement.type === 'function'
    ) {
        children.push(inst._renderedComponent);
    }

    const childrenArray = compact(children.map(n => instToJson(n)));

    return {
        type,
        props,
        children: childrenArray.length ? childrenArray : null,
        $$typeof: Symbol.for('react.test.json'),
    };
}

export function mountToJson(wrapper) {
    return instToJson(wrapper.node);
}
