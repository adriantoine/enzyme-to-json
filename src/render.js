import compact from 'lodash.compact';

const renderChildToJson = child => {
    if(!child) return;

    if(child.type === 'tag') {
        return {
            type: child.name,
            props: child.attribs,
            children: child.children && child.children.length ? compact(child.children.map(renderChildToJson)) : null,
            $$typeof: Symbol.for('react.test.json'),
        };
    } else if(child.type === 'text') {
        return child.data;
    }
};

export default wrapper => renderChildToJson(wrapper.children()[0]);
