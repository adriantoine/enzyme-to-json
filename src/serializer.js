import ShallowWrapper from 'enzyme/build/ShallowWrapper';
import ReactWrapper from 'enzyme/build/ReactWrapper';

import toJson from './';

module.exports = {
    test(wrapper) {
        return wrapper instanceof ShallowWrapper || wrapper instanceof ReactWrapper || wrapper.cheerio;
    },
    print(wrapper) {
        return toJson(wrapper);
    },
};
