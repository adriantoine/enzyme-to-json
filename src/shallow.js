import compact from 'lodash.compact';
import isPlainObject from 'lodash.isplainobject';
import omit from 'lodash.omit';
import entries from 'object.entries';
import {propsOfNode} from 'enzyme/build/Utils';
import {typeName} from 'enzyme/build/Debug';
import {childrenOfNode} from 'enzyme/build/ShallowTraversal';
import {isElement} from 'enzyme/build/react-compat';

function nodeToJson(node) {
    if(!node) return node;

    if (typeof node === 'string' || typeof node === 'number') {
        return node;
    }

    if (!isElement(node)) {
        if (Array.isArray(node)) {
            return node.map(nodeToJson);
        }

        if (isPlainObject(node)) {
            return entries(node).reduce((obj, [key, val]) => {
                obj[key] = nodeToJson(val);
                return obj;
            }, {});
        }

        return node;
    }

    const children = compact(childrenOfNode(node).map(n => nodeToJson(n)));
    const type = typeName(node);
    const props = omit(propsOfNode(node), 'children');

    return {
        type,
        props: nodeToJson(props),
        children: children.length ? children : null,
        $$typeof: Symbol.for('react.test.json'),
    };
}

export function shallowToJson(wrapper) {
    return nodeToJson(wrapper.node);
}
