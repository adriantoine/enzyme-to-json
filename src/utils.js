import ShallowWrapper from 'enzyme/build/ShallowWrapper';
import ReactWrapper from 'enzyme/build/ReactWrapper';

const SHALLOW_WRAPPER_NAME = ShallowWrapper.name;
const REACT_WRAPPER_NAME = ReactWrapper.name;

export const isShallowWrapper = wrapper => wrapper && wrapper.constructor && wrapper.constructor.name === SHALLOW_WRAPPER_NAME;
export const isReactWrapper = wrapper => wrapper && wrapper.constructor && wrapper.constructor.name === REACT_WRAPPER_NAME;
export const isCheerioWrapper = wrapper => wrapper.cheerio;

export const isEnzymeWrapper = wrapper => isShallowWrapper(wrapper) || isReactWrapper(wrapper) || isCheerioWrapper(wrapper);
