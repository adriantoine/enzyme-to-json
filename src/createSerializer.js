import {isEnzymeWrapper} from './utils';
import toJson from './';

export default function createSerializer(options) {
  return {
    test(wrapper) {
      return isEnzymeWrapper(wrapper);
    },
    print(wrapper, serializer) {
      return serializer(toJson(wrapper, options));
    },
  };
}
