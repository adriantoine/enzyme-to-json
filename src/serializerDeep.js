import {isEnzymeWrapper} from './utils';
import toJson from './';

export function test(wrapper) {
  return isEnzymeWrapper(wrapper);
}

export function print(wrapper, serializer) {
  return serializer(toJson(wrapper, {mode: 'deep'}));
}
