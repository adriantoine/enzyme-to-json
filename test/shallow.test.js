/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from '../src';
import { BasicPure } from './fixtures/pure-function';
import { BasicClass, ClassWithPure, ClassWithNull } from './fixtures/class';

function WrapperComponent(props) {
    return <BasicPure {...props} />;
}

it('converts basic pure shallow', () => {
    const shallowed = shallow(
        <BasicPure className="pure"><strong>Hello!</strong></BasicPure>
    );

    expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('converts basic class shallow', () => {
    const shallowed = shallow(
        <BasicClass className="class"><strong>Hello!</strong></BasicClass>
    );

    expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('converts a class mount with a pure function in it', () => {
    const shallowed = shallow(
        <ClassWithPure className="class"><strong>Hello!</strong></ClassWithPure>
    );
    expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('handles a component which returns null', () => {
    const shallowed = shallow(
        <ClassWithNull />
    );
    expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('handles elements in props', () => {
    const shallowed = shallow(
        <WrapperComponent element={<BasicPure><strong>Hello!</strong></BasicPure>} />
    );
    expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('handles elements in prop arrays', () => {
    const shallowed = shallow(
        <WrapperComponent elements={[
            <BasicPure><strong>Hello!</strong></BasicPure>,
        ]} />
    );
    expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('handles elements in prop objects', () => {
    const shallowed = shallow(
        <WrapperComponent element={{
            element: <BasicPure><strong>Hello!</strong></BasicPure>,
            nestedElements: [
                <BasicPure><strong>Hello again!</strong></BasicPure>,
            ],
        }} />
    );

    expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('ignores non-plain objects', () => {
    function TestConstructor() {
        this._test = true;
    }

    const shallowed = shallow(
        <WrapperComponent instance={new TestConstructor()} />
    );

    expect(shallowToJson(shallowed)).toMatchSnapshot();
});
