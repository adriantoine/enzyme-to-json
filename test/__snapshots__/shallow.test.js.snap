exports[`test converts a class mount with a pure function in it 1`] = `
<div
  className="class-with-pure class"
  onClick={[Function]}>
  <BasicPure
    className="nested-pure">
    <span>
      <strong>
        Hello!
      </strong>
    </span>
    <span
      className="empty" />
  </BasicPure>
</div>
`;

exports[`test converts basic class shallow 1`] = `
<div
  className="basic-class class"
  onClick={[Function]}>
  <div
    className="group"
    id="group-id">
    <span>
      <strong>
        Hello!
      </strong>
    </span>
    <span
      className="empty" />
  </div>
</div>
`;

exports[`test converts basic pure shallow 1`] = `
<div
  className="basic-pure pure"
  onClick={[Function]}>
  <div
    className="group"
    id="group-id">
    <span>
      <strong>
        Hello!
      </strong>
    </span>
    <span
      className="empty" />
  </div>
</div>
`;

exports[`test handles a component which returns null 1`] = `null`;

exports[`test handles elements in prop arrays 1`] = `
<BasicPure
  elements={
    Array [
      <BasicPure>
        <strong>
            Hello!
        </strong>
    </BasicPure>,
    ]
  } />
`;

exports[`test handles elements in prop objects 1`] = `
<BasicPure
  element={
    Object {
      "element": <BasicPure>
        <strong>
            Hello!
        </strong>
    </BasicPure>,
      "nestedElements": Array [
        <BasicPure>
          <strong>
                Hello again!
          </strong>
    </BasicPure>,
      ],
    }
  } />
`;

exports[`test handles elements in props 1`] = `
<BasicPure
  element={
    <BasicPure>
      <strong>
        Hello!
      </strong>
    </BasicPure>
  } />
`;

exports[`test ignores non-plain objects 1`] = `
<BasicPure
  instance={
    TestConstructor {
      "_test": true,
    }
  } />
`;
